plugin/contrib/mesos/pkg/executor/executor_test.go

```diff
@@ -195,3 +214,363 @@ func TestSuicide_WithTasks(t *testing.T) {
+							Protocol:      api.ProtocolTCP,
+						},
+					},
+				},
+			},
+		},
+		Status: api.PodStatus{
+			PodIP: fmt.Sprintf("1.2.3.%d", 4+i),
+			Conditions: []api.PodCondition{
+				{
+					Type:   api.PodReady,
+					Status: api.ConditionTrue,
+				},
+			},
+		},
+	}
+}
+
+// TODO(tyler) remove when scheduler tests are merged in, use that version
+// Create mock of pods ListWatch, usually listening on the apiserver pods watch endpoint
+type MockPodsListWatch struct {
+	ListWatch   cache.ListWatch
+	fakeWatcher *watch.FakeWatcher
+	list        api.PodList
+}
+
+// TODO(tyler) remove when scheduler tests are merged in, use that version
+// A apiserver mock which partially mocks the pods API
+type TestServer struct {
+	server *httptest.Server
+	Stats  map[string]uint
+	lock   sync.Mutex
+}
+
+// TODO(tyler) remove when scheduler tests are merged in, use that version
+func NewTestServer(t *testing.T, namespace string, pods *api.PodList) *TestServer {
+	ts := TestServer{
+		Stats: map[string]uint{},
+	}
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(testapi.ResourcePath("bindings", namespace, ""), func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	})
+
+	ts.server = httptest.NewServer(mux)
+	return &ts
+}
+
+// TODO(tyler) remove when scheduler tests are merged in, use that version
+func NewMockPodsListWatch(initialPodList api.PodList) *MockPodsListWatch {
+	lw := MockPodsListWatch{
+		fakeWatcher: watch.NewFake(),
+		list:        initialPodList,
+	}
+	lw.ListWatch = cache.ListWatch{
+		WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+			return lw.fakeWatcher, nil
+		},
+		ListFunc: func() (runtime.Object, error) {
+			return &lw.list, nil
+		},
+	}
+	return &lw
+}
+
+// TestExecutorShutdown ensures that the executor properly shuts down
+// when Shutdown is called.
+func TestExecutorShutdown(t *testing.T) {
+	mockDriver := MockExecutorDriver{}
+	kubeletFinished := make(chan struct{})
+	var exitCalled int32 = 0
+	config := Config{
+		Docker:  dockertools.ConnectToDockerOrDie("fake://"),
+		Updates: make(chan interface{}, 1024),
+		ShutdownAlert: func() {
+			close(kubeletFinished)
+		},
+		KubeletFinished: kubeletFinished,
+		ExitFunc: func(_ int) {
+			atomic.AddInt32(&exitCalled, 1)
+		},
+	}
+	executor := New(config)
+
+	executor.Init(mockDriver)
+	executor.Registered(mockDriver, nil, nil, nil)
+
+	mockDriver.On("Stop").Return(mesosproto.Status_DRIVER_STOPPED, nil).Once()
+
+	executor.Shutdown(mockDriver)
+
+	assert.Equal(t, false, executor.isConnected(),
+		"executor should not be connected after Shutdown")
+	assert.Equal(t, true, executor.isDone(),
+		"executor should be in Done state after Shutdown")
+
+	doneChanWorked := false
+	select {
+	case _, ok := <-executor.Done():
+		if !ok {
+			doneChanWorked = true
+		}
+	default:
+	}
+	assert.Equal(t, true, doneChanWorked,
+		"done channel should be closed after shutdown")
+
+	assert.Equal(t, true, atomic.LoadInt32(&exitCalled) > 0,
```

Comment:
During tests doShutdown will call it twice in the test because the process doesn't exit before the doShutdown method returns, triggering the deferred unclean shutdown func.